stdlib/map.ml

```diff
@@ -356,4 +360,42 @@ module Make(Ord: OrderedType) = struct
 
     let choose = min_binding
 
+    type 'a cursor_action =
+      | Act_yield of key * 'a
+      | Act_explore of 'a t
+
+    type 'a cursor = {
+      low: key option;
+      high: key option;
+      stack: 'a cursor_action list;
+    }
```

Comment:
I'm saying that it's unclear the range version are really that useful (you can always use `split` to get the range-restricted map in O(log n) if you want anyway), and I would not make the common-case more complex (to review, test etc.) and slightly slower because of them. You should implement what you need to be in the core to add other stuff on top of that; what you need is a `cursor_start_form key` that is passed a key and starts at this key (because you can't implement that efficiently from `cursor_start`), the rest doesn't need to be in the stdlib (for example you can implement your own `cursor` type as a triple/record on top of it if you want to provide the cursor interface you are currently proposing, or implement a `cursor_next_below key`, etc.).